gateway-service/src/main/java/org/zowe/apiml/gateway/security/service/schema/source/DefaultAuthSourceService.java

```diff
@@ -88,21 +97,55 @@ public DefaultAuthSourceService(@Autowired JwtAuthSourceService jwtAuthSourceSer
     public Optional<AuthSource> getAuthSourceFromRequest() {
         AuthSourceService service = getService(AuthSourceType.JWT);
         Optional<AuthSource> authSource = service.getAuthSourceFromRequest();
+        authSource = verifyTokenOrigin(authSource);
         if (!authSource.isPresent() && isPATEnabled) {
+            // get PAT token from PAT specific cookie or header
             service = getService(AuthSourceType.PAT);
             authSource = service.getAuthSourceFromRequest();
         }
-        if (!authSource.isPresent() && isOIDCEnabled) {
-            service = getService(AuthSourceType.OIDC);
-            authSource = service.getAuthSourceFromRequest();
-        }
-        if (!authSource.isPresent()) {
+        if (!authSource.isPresent() && isX509Enabled) {
             service = getService(AuthSourceType.CLIENT_CERT);
             authSource = service.getAuthSourceFromRequest();
         }
         return authSource;
     }
 
+    private Optional<AuthSource> verifyTokenOrigin(Optional<AuthSource> authSource) {
+        if (authSource.isPresent()) {
+            AuthSourceType currentType = authSource.get().getType();
+            String token = (String) authSource.get().getRawSource();
+            Claims claims = getJwtClaims(token);
+            QueryResponse.Source source = QueryResponse.Source.valueByIssuer(claims.getIssuer());
+            switch (source) {
+                case ZOWE:
```

Comment:
ZOWE case is empty? if yes we can remove it I guess from the switch

Reply:
ZOWE and ZOSMF origin is common for JwtAuthSource so the switch case is common for both of them.